ciao-controller/internal/datastore/datastore.go

```diff
@@ -1406,6 +1406,29 @@ func (ds *Datastore) AddBlockDevice(device types.BlockData) error {
 	return nil
 }
 
+func (ds *Datastore) DeleteBlockDevice(ID string) error {
+	// lock both tenants and devices maps
+	var err error
+
+	ds.bdLock.Lock()
+	ds.tenantsLock.Lock()
+
+	dev, ok := ds.blockDevices[ID]
+	if ok {
+		delete(ds.blockDevices, ID)
+		delete(ds.tenants[dev.TenantID].devices, ID)
+
+		go ds.db.deleteBlockData(ID)
```

Comment:
deleteBlockData() locks the sqliteDB so that can't have concurrent accesses.  But there's no relation between that and the bdLock (protecting blockDevices[]?) and tenantsLock (protecting tenants[]?).

The way it's coded here it feels like the deleteBlockData() call and its locking are meant to be hierarchically below the bdLock and tenantsLock, ie: happen while those are also held.  That would insure there's no racing.  But since it's a go routine, it's possible for the go ds.db.deleteBlockData(ID) to very minimally "start", and DeleteBlockDevice() run to completion and unlock tenantsLock and bdLock and return.

It may be then that other tenant and block device adds/removes happen.

And only later the deleteBlockData() happens, perhaps for a tenant or block device which is now removed.  I _think_ looking at sqlite3db.go that this is actually ok today.  But this is the type of things that leads to bugs later.

I suggest either adding a wait group or channel in the deleteBlockData() flow so this caller can await its completion, or moving the unlock calls upward so a future person looking at this code sees the db deletion is independent of any attempt to maintain coherency on tenants[] and blockdevices[], eg:

    	if ok {
    		delete(ds.blockDevices, ID)
    		delete(ds.tenants[dev.TenantID].devices, ID)
    		ds.tenantsLock.Unlock()
    		ds.bdLock.Unlock()
    
    		go ds.db.deleteBlockData(ID)
    		return err
    	}
    
       	ds.tenantsLock.Unlock()
       	ds.bdLock.Unlock()
    
    	return ErrNoBlockData